File: app/sequence/service.py
import os, datetime, time
from library.db_connection_factory import get_collection
import library.db_utils as db_utils

domain = 'sequence'

def is_present(space_id, field, context):
    if len(db_utils.find(space_id, domain, {'field': field, 'context':context})) == 1:
        return True
    return False

def create_sequence(space_id, field, context, initial_value, factor):
    db_utils.upsert(space_id, domain, {
        'field': field,
        'context': context,
        'nextVal': initial_value,
        'factor': factor
    })

def reset_sequence(space_id, field, context):
    sequence = db_utils.find(space_id, domain, {'field': field, 'context': context})[0]
    sequence['nextVal'] = 1
    return db_utils.upsert(space_id, domain, sequence)

def nextval(space_id, field, context):
    print(space_id, field, context)
    data = db_utils.find(space_id, domain, {'field': field, 'context': context})[0]
    currVal = data['nextVal']
    data['nextVal'] = data['nextVal'] + data['factor']
    print(data)
    db_utils.upsert(space_id, domain, data)
    return currVal

File: src/modules/sequence/service.ts
const { sequenceCollection, sequenceSchema } = require("./model");
const { getGlobalCollection, getCollection } = require("../../lib/dbutils");

export const createSequence = async (data: {
  field: string;
  context?: string;
  factor?: number;
  realm?: number;
  nextval?: number;
}) => {
  const factor = data.factor || 1;
  const context = data.context || "na";
  const realm = data.realm || 100;
  const nextval = data.nextval || 1;
  let model;
  if (realm) {
    model = getCollection(realm, sequenceCollection, sequenceSchema);
  } else {
    model = getGlobalCollection(sequenceCollection, sequenceSchema);
  }

  const existingSequence = await model.findOne({
    field: data.field,
    context,
  });

  if (existingSequence) {
    return existingSequence;
  }

  return await model.findOneAndUpdate(
    { field: data.field, context },
    {
      field: data.field,
      context,
      factor,
      nextval,
    },
    { upsert: true, new: true }
  );
};

export const nextval = async (data: {
  field: string;
  context?: string;
  realm?: number;
}) => {
  const context = data.context || "na";
  const realm = data.realm || 100;
  let model;
  if (realm) {
    model = getCollection(realm, sequenceCollection, sequenceSchema);
  } else {
    model = getGlobalCollection(sequenceCollection, sequenceSchema);
  }
  const sequence = await model.findOne({
    field: data.field,
    context,
  });
  if (!sequence) {
    return null;
  }
  await model.findOneAndUpdate(
    { field: data.field, context },
    { nextval: sequence.nextval + sequence.factor },
    { upsert: true, new: true }
  );
  return sequence.nextval;
};
